neuralprophet/forecaster.py

```diff
@@ -3288,12 +3296,30 @@ def _reshape_raw_predictions_to_forecst_df(self, df, predicted, components, pred
                     forecast_rest = components[comp][1:, self.n_forecasts - 1, j]
                     yhat = np.concatenate(([np.NaN] * self.max_lags, forecast_0, forecast_rest))
                     if prediction_frequency is not None:
-                        forecast_rest = components[comp][1:, :, j]
-                        yhat = np.concatenate(([np.NaN] * self.max_lags, forecast_0, forecast_rest.flatten()))
-                        if len(yhat) < len(df_forecast):
-                            yhat = np.concatenate((yhat, [np.NaN] * (len(df_forecast) - len(yhat))))
-                        else:
-                            yhat = yhat[: len(df_forecast)]
+                        for key, value in prediction_frequency.items():
+                            if key == "daily":
+                                dates_comp = dates[dates.dt.hour == value]
+                            elif key == "weekly":
+                                dates_comp = dates[dates.dt.dayofweek == value]
+                            elif key == "monthly":
+                                dates_comp = dates[dates.dt.day == value]
+                            elif key == "yearly":
+                                dates_comp = dates[dates.dt.month == value]
+                            elif key == "hourly":
+                                dates_comp = dates[dates.dt.minute == value]
+                            else:
+                                raise ValueError(f"prediction_frequency {key} not supported")
+                            ser = pd.Series()
+                            for date in dates_comp:
+                                d = pd.date_range(date, periods=self.n_forecasts + 1, freq=self.data_freq)
+                                ser = pd.concat((ser, pd.Series(d).iloc[1:]))
+                            df_comp = pd.DataFrame({"ds": ser, "yhat": components[comp].flatten()}).drop_duplicates(
+                                subset="ds"
+                            )
+                            df_comp, _ = df_utils.add_missing_dates_nan(df_comp, freq=self.data_freq)
+                            yhat = pd.merge(df_forecast.filter(["ds", "ID"]), df_comp, on="ds", how="left")[
+                                "yhat"
+                            ].values
```

Comment:
For any case (n_forecasts <,>,= prediction_frequency) the non-lagged components are now returned in a clean way (dropping duplicate trend predictions, or adding missing NaNs in case of n_forecasts > prediction_freq, and n_forecasts < prediction_freq, respectively)